### ISSUE TITLE
Box: "Enterprise" box_sub type does not refresh token

### ISSUE THREAD

**Issue by davidhewitt** [rclone/rclone#4091]
#### What is your current rclone version (output from `rclone version`)?

rclone v1.51.0

#### What problem are you are trying to solve?

I have a Box app, with a `config.json` downloaded for that app.

I can authenticate with `rclone config`, setting the `box_sub_type` to `enterprise` and providing the path to the config.json. `rclone` successfully authenticates and stores a config entry similar to the below:

```
[box]
type = box
box_config_file = config.json
box_sub_type = enterprise
token = {"access_token": "xyz...", "token_type": "bearer", "expiry": "<timestamp>"}
```

`rclone` will then operate correctly until after token `expiry`, from which point `rclone` fails with errors like `token expired and refresh token is not set`

#### How do you think rclone should be changed to solve that?

All information to re-authenticate is provided in `config.json`, so `rclone` should be able to generate a new token and proceed operation.

**Comment by ncw** [comment]
I don't know very much about this...

@dbramwell you contributed the original code for this - is it working OK for you?

**Comment by dbramwell** [comment]
When using the box config.json we don't get a refresh token back, so the usual way of refreshing doesn't work. We've been getting round it by automating the rclone config step as well and running that before each rclone box command.

Having rclone automatically ask for a new token using the config.json seems reasonable to me. If someone then removed the config file then rclone would be unable to get a new token but I guess that's fine. Another option would be for rclone to store the contents of the config.json, rather than its location? Not sure if that could be considered a security problem

**Comment by ncw** [comment]
The box backend knows when the token expires - this bit of code gets called. We store the location of the `config.json` in the configuration file so we should have everything we need to auto renew.

https://github.com/rclone/rclone/blob/1af9fcbbfae66fc0de0aa44eb2dd3d8c3b185321/backend/box/box.go#L396-L400

Do you fancy having a go at this @dbramwell ?

**Comment by dbramwell** [comment]
I was having a look the other day but found it quite hard to follow. I'll give it a go but will probably be back with questions.

**Comment by ncw** [comment]
Sounds great @dbramwell ! Ping me here if you need any help - thanks.

**Comment by abrar-hnxlabs** [comment]
@dbramwell  any ETA on this fix, I am also getting this error, I could also look into the code and fix this if you havent started yet.

**Comment by dbramwell** [comment]
Hey @abrar-hnxlabs. I've not started yet, was hoping to have a go this weekend, but I'm unable to give an ETA. If you have the time then please give it a go! In the mean time, you probably already know this, but rclone config can be run without user interaction:
`rclone config create box box box_config_file /path/to/config.json`

That's what I've been using in automated scripts to refresh the token before each time I want to use it.

**Comment by abrar-hnxlabs** [comment]
thanks @dbramwell  , this is very helpful and works for now as a work around the exact cmd i had to use was this 
`rclone config create box box box_config_file /path/to/config.json box_sub_type enterprise`

we ran into another issue , where our sync job actually takes more than 1 hr , so the token does expire within 1 hr and does not renew within the sync job and exits with an error.

**Comment by dbramwell** [comment]
@abrar-hnxlabs glad it helped. Not sure what the solution is to the sync job issue, I imagine even with the fix to have rclone refresh the token you would still have this issue? I'm not exactly sure how sync works. It'd be good if the token expiry time was configurable within box

**Comment by dbramwell** [comment]
@abrar-hnxlabs I've given it a go in [pr 4107](https://github.com/rclone/rclone/pull/4107). It works, but I'm not sure it's the best implementation. Looking at the way token refreshes work I think it would fix your sync problem too

**Comment by ncw** [comment]
I've merged @dbramwell 's fix to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52

@abrar-hnxlabs and @davidhewitt  can you check it works for you? Thanks

**Comment by davidhewitt** [comment]
🎉 many thanks! Not sure I'll get another swipe at this soon, but when I do I'll let you know if I have any trouble.

**Comment by abrar-hnxlabs** [comment]
@ncw  , @dbramwell  thanks  and really appreciate the quick fix and turn around on this issue
